Adding service that allows setting of position with an position confidence indicator.

Alfiegerner opened the pull request:
This is useful for when sensors can detect the real cover position, which may have been altered outside of HA Control.

Example use case:
  Door sensor gets activated when cover is fully closed, this triggers the service call setting the cover position to 0.

The service takes:

  -- entity_id
  -- position
  -- confident - a boolean which indicates that we know the given position confidentally, which alters assumed_state to False.

The internally maintained state of confidence we have in our poistion is reset to false whenever the component triggers an action.

nagyrobi commented:
I appreciate the idea.

But the problem is that this mod alters the intended behavior of the component. It disables the arrow corresponding to the current position, which is not OK for the RF covers. This defeats the purpose for which was originally this component was adapted for.
![kép](https://user-images.githubusercontent.com/1550668/88515717-6f9ccd00-cfec-11ea-8aed-87aa088953a7.png)

With RF covers you need the arrows to be usable regardless of what's the state assumed in HA, in order to be able to manually sync them at any time. Without the requirement of installing extra sensors etc.

Alfiegerner commented:
Hi, no qualms at all if you'd rather not take it. 

It should (and there may be a bug there) keep the assumed state as True unless you specify confident as True.

So by default it shouldn't change the behaviour of the arrow buttons, and if the service is not called it should operate exactly as it did before.

nagyrobi commented:
I'd like to add this in, as long as it doesn't alter the original functionality.
And what's needed additionally, is some detailed description in the readme on how to use this with examples etc.

nagyrobi commented:
Testing.... I get the following error:
```
2020-07-27 12:01:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.139809608870496] name 'position' is not defined
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 125, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1265, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1300, in _execute_service
    await handler.func(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 208, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 454, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 595, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 485, in _handle_entity_call
    await result
  File "/config/custom_components/cover_rf_time_based/cover.py", line 189, in async_set_cover_position
    await self.set_position(position)
NameError: name 'position' is not defined
```

Alfiegerner commented:
Ok thanks, looking now.

Alfiegerner commented:
Overkeen cutting going on a tidy up, should be good now.

Fixed the assumed state bug and added some notes to README, let me know if we need anything else.

nagyrobi commented:
OK thanks. Let me do some testing.

nagyrobi commented:
Looks fine. Thank you very much.

nagyrobi commented:
Bumped release version to trigger HACS update available notification for people.

Side note: what would be even more fantastic, would be to actually change the state using the timed algorithm. So for example if current position is, say, 70, if calling the service to set the position to 0 would not make it jump to 0 directly, but start the timed movement until 0 (and still not sending out the actual command). With a similar option like confident.

Would be useful for RF covers, where the RF remote triggers the cover directly, but the RF-MQTT brigde can see the RF command in the air and with this service we could "play" the movement in HA in parallel with the cover.

Alfiegerner commented:
No problem, we can add a 'travel_to' parameter that will cause timed movememnt.

I was thinking of doing something like that for when the service is called and the cover is in motion.  Currently it stops the position at that given in the service, whereas it should probably use that as the start position for a new timed movement to the old target.

So currently works well for my main use case of sensors at 100 and 0, but could do with a tweak if people want sensors at other positions.   

Will have a look tonight.

Just a thought - would you consider a change of name for this component? I know you're using for RF but others will for other technologies (my relays are controlled by command line and GPIO pin outs), and I think the component itself is implementation agnostic and good example of a generic cover handler.

nagyrobi commented:
I've been considering a name change - need a good idea - and I'm also pondering in to submitting this whole thing to HA's main repo as a built-in component.

Alfiegerner commented:
Will have a think - it's kind of a replacement / enhancement for the template sensor, but we'd need to make the service calls more generic than specific script calls.
